File: C/src/static_stack.c
#include <stdio.h>
#include "static_stack.h"

void inicializarStaticStack(StaticStack *s) {
    s->topo = -1;
}

int pilhaVazia(StaticStack *s) {
    return s->topo == -1;
}

int pilhaCheia(StaticStack *s) {
    return s->topo == MAXTAM - 1;
}

int empilhar(StaticStack *s, Registro r) {
    if (pilhaCheia(s)) return 0;

    s->dados[++s->topo] = r;
    return 1;
}

int desempilhar(StaticStack *s, Registro *r) {
    if (pilhaVazia(s)) return 0;

    *r = s->dados[s->topo--];
    return 1;
}

void imprimirStaticStack(StaticStack *s) {
    if (pilhaVazia(s)) {
        printf("Pilha vazia.\n");
        return;
    }

    for (int i = s->topo; i >= 0; i--) {
        printf("%d ", s->dados[i].valor);
    }
    printf("\n");
}

void salvarStaticStackEmArquivo(StaticStack *s, const char *nomeArquivo) {
    FILE *arquivo = fopen(nomeArquivo, "w");
    if (arquivo == NULL) {
        perror("Erro ao abrir o arquivo para escrita");
        return;
    }

    for (int i = s->topo; i >= 0; i--) {
        fprintf(arquivo, "%d\n", s->dados[i].valor);
    }

    fclose(arquivo);
}

File: Java/src/StaticStack.java

public class StaticStack {
    protected Registro[] elementos;
    protected int topo;
    protected int capacidade;

    public StaticStack(int capacidade) {
        this.capacidade = capacidade;
        this.elementos = new Registro[capacidade];
        this.topo = -1;
    }

    public boolean empilhar(Registro valor) {
        if (this.topo == this.capacidade - 1) return false; // Pilha cheia

        this.topo++;
        this.elementos[this.topo] = valor;
        return true;
    }

    public Registro desempilhar() {
        if (pilhaVazia()) return null; // Pilha vazia

        Registro valorRemovido = this.elementos[this.topo];
        this.topo--;
        return valorRemovido;
    }

    public boolean pilhaVazia() {
        return this.topo == -1;
    }

    public void imprimir() {
        if (pilhaVazia()) {
            System.out.println("Pilha vazia.");
            return;
        }
        for (int i = this.topo; i >= 0; i--) {
            System.out.print(this.elementos[i].valor + " ");
        }
        System.out.println();
    }

    public void salvarEmArquivo(String nomeArquivo) {
        try (java.io.FileWriter writer = new java.io.FileWriter(nomeArquivo)) {
            for (int i = this.topo; i >= 0; i--) {
                writer.write(this.elementos[i].valor + "\n");
            }
        } catch (java.io.IOException e) {
            System.err.println("Erro ao abrir arquivo para escrita: " + e.getMessage());
        }
    }
}

